internal/ethapi/api.go

```diff
@@ -1232,21 +1232,25 @@ func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNr
 		if err != nil {
 			return 0, err
 		}
-		if metaTxParams != nil {
-			sponsorAmount, _ := types.CalculateSponsorPercentAmount(metaTxParams, feeCap)
-			sponsorBalance := state.GetBalance(metaTxParams.GasFeeSponsor)
-			if sponsorBalance.Cmp(sponsorAmount) <= 0 {
-				return 0, types.ErrSponsorBalanceNotEnough
-			}
-			balance = new(big.Int).Add(balance, sponsorAmount)
-		}
+
 		available := new(big.Int).Set(balance)
 		if args.Value != nil {
-			if args.Value.ToInt().Cmp(available) >= 0 {
+			if args.Value.ToInt().Cmp(available) > 0 {
 				return 0, core.ErrInsufficientFundsForTransfer
 			}
 			available.Sub(available, args.Value.ToInt())
 		}
+
+		if metaTxParams != nil {
+			sponsorAmount, _ := types.CalculateSponsorPercentAmount(metaTxParams, new(big.Int).Mul(feeCap, new(big.Int).SetUint64(hi)))
+			sponsorBalance := state.GetBalance(metaTxParams.GasFeeSponsor)
```

Comment:
```
sponsorAmount, selfAmount := types.CalculateSponsorPercentAmount(metaTxParams, new(big.Int).Mul(feeCap, new(big.Int).SetUint64(hi)))
```
`available` should be no less than `selfAmount`. Please add this check.

Reply:
`sponsorAmount` and `selfAmount` will be a large number because we get them by `CalculateSponsorPercentAmount` with input `feeCap * hi`. The `hi` is kind of blockGasLimit, a huge number.

Reply:
calculate amount inside `estimateGas` function is merely to assign a value to "available" but can not verify whether the balance is sufficient.
Or whether we can estimateGas with payload without metatx prefix `pureGasLimit`，then add a const number to the gasLimit `metaTxGasLimit`.
CalculateSponsorPercentAmount use `metaTxGasLimit` as input, `sponsorAmount & selfAmount` can be more accurate. Checking the balance would also be more meaningful.